metadata/src/validation.rs

```diff
@@ -0,0 +1,851 @@
+        .ok_or(NotFound::Pallet)?;
+
+    let call_id = pallet.calls.as_ref().ok_or(NotFound::Item)?.ty.id;
+
+    let call_ty = metadata.types.resolve(call_id).ok_or(NotFound::Item)?;
+
+    let call_variants = match &call_ty.type_def {
+        TypeDef::Variant(variant) => &variant.variants,
+        _ => return Err(NotFound::Item),
+    };
+
+    let variant = call_variants
+        .iter()
+        .find(|v| v.name == call_name)
+        .ok_or(NotFound::Item)?;
+
+    // hash the specific variant representing the call we are interested in.
+    let hash = get_variant_hash(&metadata.types, variant, &mut HashSet::new());
+    Ok(hash)
+}
+
+/// Obtain the hash representation of a `frame_metadata::v15::PalletMetadata`.
+pub fn get_pallet_hash(
+    registry: &PortableRegistry,
+    pallet: &PalletMetadata<PortableForm>,
+) -> [u8; 32] {
+    // Begin with some arbitrary hash (we don't really care what it is).
+    let mut bytes = hash(&[19]);
+    let mut visited_ids = HashSet::<u32>::new();
+
+    if let Some(calls) = &pallet.calls {
+        bytes = xor(
+            bytes,
+            get_type_hash(registry, calls.ty.id, &mut visited_ids),
+        );
+    }
+    if let Some(ref event) = pallet.event {
+        bytes = xor(
+            bytes,
+            get_type_hash(registry, event.ty.id, &mut visited_ids),
+        );
+    }
+    for constant in pallet.constants.iter() {
+        bytes = xor(bytes, hash(constant.name.as_bytes()));
+        bytes = xor(
+            bytes,
+            get_type_hash(registry, constant.ty.id, &mut visited_ids),
+        );
+    }
+    if let Some(ref error) = pallet.error {
+        bytes = xor(
+            bytes,
+            get_type_hash(registry, error.ty.id, &mut visited_ids),
+        );
+    }
+    if let Some(ref storage) = pallet.storage {
+        bytes = xor(bytes, hash(storage.prefix.as_bytes()));
+        for entry in storage.entries.iter() {
+            bytes = hash_hashes(
+                bytes,
+                get_storage_entry_hash(registry, entry, &mut visited_ids),
+            );
+        }
+    }
+
+    bytes
+}
+
+/// Obtain the hash representation of a `frame_metadata::v15::RuntimeMetadataV15`.
+pub fn get_metadata_hash(metadata: &RuntimeMetadataV15) -> [u8; 32] {
+    // Collect all pairs of (pallet name, pallet hash).
+    let mut pallets: Vec<(&str, [u8; 32])> = metadata
+        .pallets
+        .iter()
+        .map(|pallet| {
+            let hash = get_pallet_hash(&metadata.types, pallet);
+            (&*pallet.name, hash)
+        })
+        .collect();
+
+    // Sort by pallet name to create a deterministic representation of the underlying metadata.
+    pallets.sort_by_key(|&(name, _hash)| name);
+
+    // Note: pallet name is excluded from hashing.
+    // Each pallet has a hash of 32 bytes, and the vector is extended with
+    // extrinsic hash and metadata ty hash (2 * 32).
+    let mut bytes = Vec::with_capacity(pallets.len() * 32 + 64);
```

Comment:
You could also add constants for these instead such `(pallet.len() * HASH_LEN + EXTRINSIC_HASH_LEN + METADATA_HASH_LEN)` for readability

The comment is good enough anyway

Reply:
That makes sense, have added a `HASH_LEN` which should be universal across this file since it is produced by our `hash` function, thanks!